drivers/iio/dac/ad5413.c

```diff
@@ -0,0 +1,780 @@
+            return ret;
+        }
+        /* Validate current range: only 0–24000 µA supported */
+        if (tmparray[0] != 0 || tmparray[1] != 24000) {
+            dev_err(&st->spi->dev,
+                    "Invalid current range %u-%u µA; only 0–24000 supported\n",
+                    tmparray[0], tmparray[1]);
+            return -EINVAL;
+        }
+        range = ad5413_current_range;
+        size  = ARRAY_SIZE(ad5413_current_range);
+    }
+
+    /* 2) find_out_range() and st->out_range */
+    ret = ad5413_find_out_range(st, range, size,
+                                tmparray[0], tmparray[1]);
+    if (ret) {
+        dev_err(&st->spi->dev,
+                "Invalid range values (%u, %u)\n",
+                tmparray[0], tmparray[1]);
+        return ret;
+    }
+
+    /* 3) read slew-time (optional) */
+    ret = device_property_read_u32(&st->spi->dev,
+                                   "adi,slew-time-us", &tmp);
+    if (ret == -ENODATA || ret == -EINVAL) {
+        tmp = 0;
+    } else if (ret < 0) {
+        dev_err(&st->spi->dev,
+                "Failed to read adi,slew-time-us: %d\n", ret);
+        return ret;
+    }
+
+    return 0;
+}
+
+static int ad5413_init(struct ad5413_state *st)
+{
+	int regval, ret;
+
+	st->gpio_reset = devm_gpiod_get_optional(&st->spi->dev, "reset",
+						 GPIOD_OUT_HIGH);
+	if (IS_ERR(st->gpio_reset))
+		return PTR_ERR(st->gpio_reset);
+
+	/* Disable CRC checks */
+	ret = ad5413_crc_disable(st);
+	if (ret < 0)
+		return ret;
+
+	/* Perform a reset */
+	ret = ad5413_reset(st);
+	if (ret < 0)
+		return ret;
+
+	/* Disable CRC checks */
+	ret = ad5413_crc_disable(st);
+	if (ret < 0)
+		return ret;
+
+	/* Perform a calibration memory refresh */
+	ret = ad5413_calib_mem_refresh(st);
+	if (ret < 0)
+		return ret;
+
+	regval = ad5413_spi_reg_read(st, AD5413_REG_DIGITAL_DIAG_RESULTS);
+	if (regval < 0)
+		return regval;
+
+	/* Clear all the error flags */
+	ret = ad5413_spi_reg_write(st, AD5413_REG_DIGITAL_DIAG_RESULTS, regval);
+	if (ret < 0)
+		return ret;
+
+	/* Configure the output range */
+	ret = ad5413_set_out_range(st, st->out_range.reg);
+	if (ret < 0)
+		return ret;
+
+	/* Enable Slew Rate Control, set the slew rate clock and step */
+	if (st->slew_time) {
+		ret = ad5413_slew_rate_config(st);
+		if (ret < 0)
+			return ret;
+	}
+
+	/* Power up the DAC and internal (INT) amplifiers */
+	ret = ad5413_internal_buffers_en(st, 1);
+	if (ret < 0)
+		return ret;
+
+	/* Enable VIOUT */
+	return ad5413_spi_write_mask(st, AD5413_REG_DAC_CONFIG,
+				     AD5413_REG_DAC_CONFIG_OUT_EN_MSK,
+				     AD5413_REG_DAC_CONFIG_OUT_EN_MODE(1));
+}
+
+static int ad5413_probe(struct spi_device *spi)
+{
+	struct ad5413_state *st;
+	struct iio_dev *indio_dev;
+	int ret;
+
+	indio_dev = devm_iio_device_alloc(&spi->dev, sizeof(*st));
+	if (!indio_dev)
+		return -ENOMEM;
+
+	st = iio_priv(indio_dev);
+
+	st->spi = spi;
+
+	mutex_init(&st->lock);
```

Comment:
```suggestion
	ret = devm_mutex_init(&st->lock);
	if (ret)
		return ret;	
```

Not sure which kernel version this was introduced in, but is preferred in new code.

Reply:
Replaced mutex_init() with devm_mutex_init() as suggested, including error handling.